api/v1beta1/cluster_types.go

```diff
@@ -451,6 +469,77 @@ type ClusterStatus struct {
 	// ObservedGeneration is the latest generation observed by the controller.
 	// +optional
 	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// v1beta2 groups all the fields that will be added or modified in Cluster's status with the V1Beta2 version.
+	// +optional
+	V1Beta2 *ClusterV1Beta2Status `json:"v1beta2,omitempty"`
+}
+
+// ClusterV1Beta2Status groups all the fields that will be added or modified in Cluster with the V1Beta2 version.
+// See https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20240916-improve-status-in-CAPI-resources.md#machineset-newconditions for more context.
+type ClusterV1Beta2Status struct {
+	// conditions represents the observations of a Cluster's current state.
+	// Known condition types are Available, InfrastructureReady, ControlPlaneInitialized, ControlPlaneAvailable, WorkersAvailable, MachinesReady
+	// MachinesUpToDate, RemoteConnectionProbe, ScalingUp, ScalingDown, Remediating, Deleting, Paused.
+	// Additionally, a TopologyReconciled will be added in case the Cluster is referencing a ClusterClass / defining a managed Topology.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// controlPlane groups all the observations about Cluster's ControlPlane current state.
+	// +optional
+	ControlPlane *ClusterControlPlaneStatus `json:"controlPlane,omitempty"`
+
+	// workers groups all the observations about Cluster's Workers current state.
+	// +optional
+	Workers *WorkersStatus `json:"workers,omitempty"`
+}
+
+// ClusterControlPlaneStatus groups all the observations about control plane current state.
+type ClusterControlPlaneStatus struct {
+	// desiredReplicas is the total number of desired control plane machines in this cluster.
+	// +optional
+	DesiredReplicas int32 `json:"desiredReplicas"`
```

Comment:
Do we need to care about whether this has been observed deliberately or not?

Currently, since this is not a pointer and has no omitempty, when I marshal into json, I will get `desiredReplicas: 0` even if I haven't set the field. This could then cause a discrepancy with the desired replicas from the spec, and look like the controller has observed a desire for 0 replicas.

Also, should we not expect the controller to always set this field? Should it actually be required? Same probably applies to lots of the other fields as well

Reply:
Made all pointers,
We have to do the same for all the existing counters when we move to v1beta2 (will add to the umbrella issue)